src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -45,127 +53,135 @@ namespace Microsoft.VisualStudio.LanguageServices.Packaging
-        // project.  If we hear about a solution level change, we'll refresh all projects.
-        private bool _solutionChanged;
-        private readonly HashSet<ProjectId> _changedProjects = new HashSet<ProjectId>();
+        /// <summary>
+        /// Used to keep track of what types of changes we've seen so we can then determine what to refresh on the UI
+        /// thread.  If we hear about project changes, we only refresh that project.  If we hear about a solution level
+        /// change, we'll refresh all projects.
+        /// </summary>
+        /// <remarks>
+        /// <c>solutionChanged == true iff changedProject == null</c> and <c>solutionChanged == false iff changedProject
+        /// != null</c>. So technically having both values is redundant.  However, i like the clarity of having both.
+        /// </remarks>
+        private readonly AsyncBatchingWorkQueue<(bool solutionChanged, ProjectId? changedProject)>? _workQueue;
 
         private readonly ConcurrentDictionary<ProjectId, ProjectState> _projectToInstalledPackageAndVersion =
             new ConcurrentDictionary<ProjectId, ProjectState>();
 
+        /// <summary>
+        /// Lock used to protect reads and writes of <see cref="_packageSourcesTask"/>.
+        /// </summary>
+        private readonly object _gate = new object();
+
+        /// <summary>
+        /// Task uses to compute the set of package sources on demand when asked the first time.  The value will be
+        /// computed and cached in the task.  When this value changes, the task will simply be cleared out.
+        /// </summary>
+        private Task<ImmutableArray<PackageSource>>? _packageSourcesTask;
+
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public PackageInstallerService(
             IThreadingContext threadingContext,
+            IAsynchronousOperationListenerProvider listenerProvider,
             VisualStudioWorkspaceImpl workspace,
             SVsServiceProvider serviceProvider,
+            [Import("Microsoft.VisualStudio.Shell.Interop.SAsyncServiceProvider")] object asyncServiceProvider,
             IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
-            [Import(AllowDefault = true)] Lazy<IVsPackageInstallerServices> packageInstallerServices,
-            [Import(AllowDefault = true)] Lazy<IVsPackageInstaller2> packageInstaller,
-            [Import(AllowDefault = true)] Lazy<IVsPackageUninstaller> packageUninstaller,
-            [Import(AllowDefault = true)] Lazy<IVsPackageSourceProvider> packageSourceProvider)
-            : base(threadingContext, workspace, SymbolSearchOptions.Enabled,
-                              SymbolSearchOptions.SuggestForTypesInReferenceAssemblies,
-                              SymbolSearchOptions.SuggestForTypesInNuGetPackages)
+            [Import(AllowDefault = true)] Lazy<IVsPackageInstallerServices>? packageInstallerServices,
```

Comment:
Did we confirm that if the import isn't satisfied that the Lazy itself is null, or the Lazy is not null and .Value is null? For some reason I thought the ? would be inside the type argument but you now have me thinking I'm wrong and it looks like we don't have other examples of this in Roslyn. An easy test would be to add a test import to this method importing some obviously-not-MEF-exported type and you can see what you get.

Reply:
(Hilariously the docs don't say which way this is...)

Reply:
i got a null param there.  so this approach is correct :)